In [1]:
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore')



In [2]:
df = pd.read_csv ('Dataset//uk_accident_data.csv')


In [3]:
df


Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious    05/06/2019  51.506187   
1       200701BS65737           Serious    02/07/2019  51.495029   
2       200701BS66127           Serious    26/08/2019  51.517715   
3       200701BS66128           Serious    16/08/2019  51.495478   
4       200701BS66837            Slight    03/09/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18/02/2022  57.374005   
660675  201091NM01881            Slight    21/02/2022  57.232273   
660676  201091NM01935            Slight    23/02/2022  57.585044   
660677  201091NM01964           Serious    23/02/2022  57.214898   
660678  201091NM02142           Serious    28/02/2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

<h1>1. How many total accidents are recorded in the dataset?</h1>

In [4]:
df.shape[0]


660679

<h1>Insight shows The dataset contains 660,679 accidents.</h1>

<h1>2. How many accidents were classified as Fatal?</h1>

In [5]:
(df['Accident_Severity'] == 'Fatal').sum()


np.int64(8661)

<h1>Insight shows the number of Fatal accidents </h1>

<h1>3. Which district has the highest average casualties per accident?</h1>

In [9]:
df.groupby('District Area')['Number_of_Casualties'].mean().idxmax()


'Blaeu Gwent'

<h1>Insights shows its blaeu gwent</h1>

<h1>4.Is it more dangerous to drive at night?</h1>

In [14]:
(df[df['Accident_Severity']=='Fatal']['Light_Conditions']
   .str.contains('Dark', na=False)
   .mean())


np.float64(0.41392448908901974)

<h1>insights shows that OF COURSE ITS DANGEROUS TO DRIVE AT NIGHT</h1>

<h1>5. What percentage of accidents were Fatal?</h1>

In [15]:
(df['Accident_Severity'].value_counts(normalize=True) * 100).round(2)


Accident_Severity
Slight     85.34
Serious    13.35
Fatal       1.31
Name: proportion, dtype: float64

<h1>Insights shows its 1.30 or sumthing</h1>

<h1>6. Which district has the most accidents recorded?</h1>

In [17]:
df['District Area'].value_counts().idxmax()


'Birmingham'

<H1>i dont know about you but i can clearly see the insight in here</H1>

<h1>7. How many accidents occurred during Daylight vs Darkness?</h1>

In [18]:
df['Light_Conditions'].value_counts()


Light_Conditions
Daylight                       484880
Darkness - lights lit          129335
Darkness - no lighting          37437
Darkness - lighting unknown      6484
Darkness - lights unlit          2543
Name: count, dtype: int64

<h1>Insights shows the stark difference between day and night accidnets</h1>

<h1>8. What’s the most common weather condition during accidents?</h1>

In [19]:
df['Weather_Conditions'].mode()[0]


'Fine no high winds'

<h1>Insights shows that people get complecent and get into more accidnets if the weather is good</h1>

<H1>9. How many accidents occurred in Urban vs Rural areas?</H1>

In [20]:
df['Urban_or_Rural_Area'].value_counts()


Urban_or_Rural_Area
Urban          421663
Rural          238990
Unallocated        11
Name: count, dtype: int64

<H1>Insights shows that urban areas have more accidents</H1>

<h1>10. What is the maximum number of vehicles involved in a single accident?</h1>

In [21]:
df['Number_of_Vehicles'].max()


np.int64(32)

<h1>11. Do urban or rural areas have higher casualties per accident?</h1>

In [22]:
df.groupby('Urban_or_Rural_Area')['Number_of_Casualties'].mean()


Urban_or_Rural_Area
Rural          1.479204
Unallocated    1.181818
Urban          1.287815
Name: Number_of_Casualties, dtype: float64

<h1>Insights shows that yes it is</h1>
<h4> i know the question looks the same as 9 but pls gib me credit theres only so muhc question i can think of at night :<</h4>

<h1>12 What’s the most dangerous vehicle type in terms of average casualties per accident?</h1>

In [23]:
df.groupby('Vehicle_Type')['Number_of_Casualties'].mean().sort_values(ascending=False).head(5)


Vehicle_Type
Data missing or out of range       1.500000
Pedal cycle                        1.370558
Taxi/Private hire car              1.368663
Goods over 3.5t. and under 7.5t    1.362861
Motorcycle over 500cc              1.359434
Name: Number_of_Casualties, dtype: float64

<h1>Insights shows its......idk???? its not showing it says it cant show </h1>

<h1>13. Does more vehicles involve mean more casualties?</h1>

In [26]:
df.groupby('Number_of_Vehicles')['Number_of_Casualties'].mean()


Number_of_Vehicles
1      1.170932
2      1.374880
3      1.711169
4      1.995575
5      2.315341
6      2.612137
7      3.064189
8      3.401361
9      3.350877
10     3.629630
11     4.000000
12     2.285714
13     7.833333
14     5.444444
15     5.000000
16     6.000000
19    13.000000
28    16.000000
32     5.000000
Name: Number_of_Casualties, dtype: float64

<h1>Insight shows maybe? i think thats a yes </h1>

<h1>14. Which month has the highest average casualties per accident?</h1>

In [30]:
df["Accident Date"] = pd.to_datetime(df["Accident Date"], dayfirst=True, errors="coerce")
df["Month"] = df["Accident Date"].dt.month_name()
monthly_avg = df.groupby("Month")["Number_of_Casualties"].mean().sort_values(ascending=False)
monthly_avg


Month
August       1.396324
July         1.366350
April        1.364796
December     1.362972
May          1.362791
March        1.351107
June         1.350826
November     1.348024
October      1.347801
September    1.347144
February     1.343941
January      1.343585
Name: Number_of_Casualties, dtype: float64

<h1>insights shows that The month with the highest average casualties per accident is August, with an average of about 1.40 casualties per accident. </h1>
<h4>MY GOD THAT TOOK "TOO" LONG TO CODE </h4>

<h1>15. Which road surface condition is most correlated with fatal outcomes?</h1>

In [32]:
fatal_df = df[df["Accident_Severity"] == "Fatal"]
fatal_counts = fatal_df["Road_Surface_Conditions"].value_counts(normalize=True)
overall_counts = df["Road_Surface_Conditions"].value_counts(normalize=True)
fatal_counts, overall_counts


(Road_Surface_Conditions
 Dry                     0.668437
 Wet or damp             0.302575
 Frost or ice            0.022289
 Snow                    0.004042
 Flood over 3cm. deep    0.002656
 Name: proportion, dtype: float64,
 Road_Surface_Conditions
 Dry                     0.678565
 Wet or damp             0.282911
 Frost or ice            0.028058
 Snow                    0.008925
 Flood over 3cm. deep    0.001541
 Name: proportion, dtype: float64)

<h1> Insights shows that its Dry road which is weird considering thats the best road condition</h1>

<h1>16. What is the average number of casualties per accident?</h1>

In [33]:
df['Number_of_Casualties'].mean().round(2)


np.float64(1.36)

<h1>insiights shows that its around 1.36</h1>

<h1>17. What is the correlation between the number of casualties and the number of vehicles involved in accidents?</h1>

In [34]:
df['Number_of_Casualties'].corr(df['Number_of_Vehicles'])


np.float64(0.22888886126927563)

<h1>insights shows more vehicles usually mean more casualties, but not very strongly.</h1>

<h1>18. What is the correlation between latitude and longitude?</h1>

In [35]:
df['Latitude'].corr(df['Longitude'])


np.float64(-0.3981136594753893)


<h1> insights as latitude increases, longitude tends to decrease </h1>

<h1>19. What is the correlation between latitude and the number of casualties?</h1>

In [36]:
df['Latitude'].corr(df['Number_of_Casualties'])


np.float64(0.032200913378368654)

<h1>insights shows its weak meanign it has almost no correlation with the amount of casualties </h1>

<h1>20. What is the correlation between longitude and the number of casualties?</h1>

In [37]:
df['Longitude'].corr(df['Number_of_Casualties'])


np.float64(-0.04040593604880228)

<h1>insights shows its the same with lattitude meanign it has almost no correlation</h1>